src/QUICClient.ts

```diff
@@ -223,233 +212,405 @@ class QUICClient extends EventTarget {
+        // Remove the stopped listener, as we intend to stop the socket
+        this.socket.removeEventListener(
+          events.EventQUICSocketStopped.name,
+          this.handleEventQUICSocketStopped
         );
+        try {
+          // Force stop of the socket even if it had a connection map
+          // This is because we will be stopping this `QUICServer` which
+          // which will stop all the relevant connections
+          await this.socket.stop({ force: true });
+        } catch (e) {
+          // Caller error would mean a domain error here
+          const e_ = new errors.ErrorQUICClientInternal(
+            'Failed to stop QUICSocket',
+            { cause: e }
+          );
+          this.dispatchEvent(
+            new events.EventQUICClientError({ detail: e_ })
+          );
+        }
       }
-    } else {
-      this.dispatchEvent(event);
+    }
+    this._closed = true;
+    this.resolveClosedP();
+    if (!this[destroyed] && this[status] !== 'destroying') {
+      // Failing this is a software error
+      await this.destroy({ force: true });
+    }
+  };
+
+  // This should only be done if it is encapsulated
+  protected handleEventQUICSocket = (evt: EventAll) => {
+    if (evt.detail instanceof AbstractEvent) {
+      this.dispatchEvent(evt.detail.clone());
     }
   };
 
   /**
-   * This must not throw any exceptions.
+   * Attached once
    */
-  protected handleQUICConnectionEvents = async (
-    event: events.QUICConnectionEvent,
-  ) => {
-    if (event instanceof events.QUICConnectionErrorEvent) {
-      this.dispatchEvent(
-        new events.QUICClientErrorEvent({
-          detail: new errors.ErrorQUICClient('Connection error', {
-            cause: event.detail,
-          }),
-        }),
+  protected handleEventQUICSocketStopped = () => {
+    const e = new errors.ErrorQUICClientSocketNotRunning();
+    this.removeEventListener(
+      EventAll.name,
+      this.handleEventQUICSocket
+    );
+    this.dispatchEvent(
+      new events.EventQUICClientError({
+        detail: e,
+      }),
+    );
+  };
+
+  protected handleEventQUICConnection = (evt: EventAll) => {
+    if (evt.detail instanceof AbstractEvent) {
+      this.dispatchEvent(evt.detail.clone());
+    }
+  };
+
+  /**
+   * All connection errors are redispatched as client errors.
+   * Connecition errors encompass both graceful closes and non-graceful closes.
+   * This also includes `ErrorQUICConnectionInternal`
+   */
+  protected handleEventQUICConnectionError = (evt: events.EventQUICConnectionError) => {
+    const error = evt.detail;
+    this.dispatchEvent(new events.EventQUICClientError(
+      { detail: error }
+    ));
+  };
+
+  /**
+   * This must be attached multiple times.
+   */
+  protected handleEventQUICConnectionSend = async (evt: events.EventQUICConnectionSend) => {
+    try {
+      await this.socket.send_(
+        evt.detail.msg,
+        evt.detail.port,
+        evt.detail.address,
+      );
```

Comment:
New race condition. If the socket stops first then this is getting triggered by the connection stopping.

```
INFO:QUICClient Test.QUICServer.QUICConnection 83668673c4c533d4fb56f9511ce7fb9ad50cf870:Started QUICConnection
INFO:QUICClient Test.QUICClient.QUICSocket:Stop QUICSocket on 127.0.0.1:40591
INFO:QUICClient Test.QUICClient.QUICSocket:Stopped QUICSocket on 127.0.0.1:40591
ERROR:QUICClient Test.QUICClient:ErrorQUICClientSocketNotRunning: QUIC Client cannot be created with an unstarted shared QUIC socket - 
INFO:QUICClient Test.QUICClient.QUICConnection ba5d0b60f693f63ab67d1cc579d43fe70d310a04:Stop QUICConnection
ERROR:QUICClient Test.QUICClient.QUICConnection ba5d0b60f693f63ab67d1cc579d43fe70d310a04:ErrorQUICConnectionLocal: QUIC Connection local error - Locally closed with application code 1
  ERROR:QUICClient Test.QUICClient:ErrorQUICClientInternal: QUIC Client internal error - Failed to send data on the QUICSocket
console.error
    ErrorQUICSocketNotRunning: 
        at constructor_.send_ (/home/faulkes/matrixcode/polykey/js-quic/src/QUICClient.ts:457:25)
        at EventTarget.handler (/home/faulkes/matrixcode/polykey/js-quic/node_modules/@matrixai/events/src/Evented.ts:121:35)
        at EventTarget.[nodejs.internal.kHybridDispatch] (node:internal/event_target:778:20)
        at EventTarget.dispatchEvent (node:internal/event_target:713:26)
        at constructor_.dispatchEvent (/home/faulkes/matrixcode/polykey/js-quic/node_modules/@matrixai/events/src/Evented.ts:285:43)
        at constructor_.dispatchEvent (/home/faulkes/matrixcode/polykey/js-quic/src/QUICConnection.ts:865:12)
        at constructor_.send (/home/faulkes/matrixcode/polykey/js-quic/node_modules/@matrixai/async-init/src/StartStop.ts:141:20)
        at constructor_.send [as handleEventQUICConnectionClose] (/home/faulkes/matrixcode/polykey/js-quic/src/QUICConnection.ts:239:18)
        at EventTarget.handler (/home/faulkes/matrixcode/polykey/js-quic/node_modules/@matrixai/events/src/Evented.ts:121:35)
        at EventTarget.[nodejs.internal.kHybridDispatch] (node:internal/event_target:778:20)
        at EventTarget.dispatchEvent (node:internal/event_target:713:26)
        at constructor_.dispatchEvent (/home/faulkes/matrixcode/polykey/js-quic/node_modules/@matrixai/events/src/Evented.ts:285:43)
        at constructor_.dispatchEvent [as handleEventQUICConnectionError] (/home/faulkes/matrixcode/polykey/js-quic/src/QUICConnection.ts:213:10)
        at EventTarget.handler (/home/faulkes/matrixcode/polykey/js-quic/node_modules/@matrixai/events/src/Evented.ts:121:35)
        at EventTarget.[nodejs.internal.kHybridDispatch] (node:internal/event_target:778:20)
        at EventTarget.dispatchEvent (node:internal/event_target:713:26)
        at constructor_.dispatchEvent (/home/faulkes/matrixcode/polykey/js-quic/node_modules/@matrixai/events/src/Evented.ts:285:43)
        at constructor_.dispatchEvent (/home/faulkes/matrixcode/polykey/js-quic/src/QUICConnection.ts:586:12)
        at /home/faulkes/matrixcode/polykey/js-quic/node_modules/@matrixai/async-init/src/StartStop.ts:99:36
        at /home/faulkes/matrixcode/polykey/js-quic/node_modules/@matrixai/async-locks/src/RWLockWriter.ts:252:64
        at withF (/home/faulkes/matrixcode/polykey/js-quic/node_modules/@matrixai/resources/src/utils.ts:24:18)
        at processTicksAndRejections (node:internal/process/task_queues:95:5)
        at constructor_.handleEventQUICClientClose (/home/faulkes/matrixcode/polykey/js-quic/src/QUICClient.ts:372:5)
        at EventTarget.handler (/home/faulkes/matrixcode/polykey/js-quic/node_modules/@matrixai/events/src/Evented.ts:136:19) {
      data: {},
      cause: undefined,
      timestamp: 2023-09-21T00:26:03.493Z
    }

      461 |       );
      462 |     } catch (e) {
    > 463 |       console.error(e);
          |               ^
      464 |       // Caller error means a domain error here
      465 |       const e_ = new errors.ErrorQUICClientInternal(
      466 |         'Failed to send data on the QUICSocket',

      at constructor_.error (src/QUICClient.ts:463:15)
      at EventTarget.handler (node_modules/@matrixai/events/src/Evented.ts:136:19)
```

Reply:
That's cause it's running asynchronously. If the socket is stopped, don't run it. Do a check condition on the running symbol.

Reply:
Do we want to check in the send handler for the server/client or in the stop method before dispatching send in connection? The streams could also be triggering their own send events.

Reply:
I think it should in the event handler. Try it and see what happens?

Reply:
I have a tempfix doing that and it's working.

Reply:
Needs to be done for the quicstreamsend handler too

Reply:
Can you push up the fix for this and quic stream send handler and this and the above one can be resolved.

Reply:
Fixed in `tmp: temp fix for socket send errors when socket ends first` for the server and client. the fix is also applied to the `QUICConnection` but there's still a bug with `QUICConnection.send()` pending an uptade to  `async-init` .

Reply:
Done